### PR TITLE
Initial implementation of Mutexes and Condition Variables

### PR DESCRIPTION
**Pull request by kayceesrk**

### CODE REVIEW

File: runtime/domain.c [hunk 1]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+  wrapper = caml_alloc_custom(&caml_mutex_ops, sizeof(caml_plat_mutex*), 0, 1);
+  Mutex_val(wrapper) = m;
+  return wrapper;
+}
+
+CAMLprim value caml_mutex_lock (value wrapper)
+{
+  CAMLparam1(wrapper); /* prevent deallocation of wrapper and destruction of
+                          mutex */
+  caml_plat_mutex* m = Mutex_val(wrapper);
```
Comment (ctk21) [inline runtime/domain.c]: We should move the roots and wrapper to the pattern used here:
 https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/otherlibs/systhreads/st_stubs.c#L738-L744
which utilizes try_lock as a fast path for an uncontended lock.

File: runtime/domain.c [hunk 2]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+  caml_plat_lock(m);
+  caml_leave_blocking_section();
+  CAMLreturn (Val_unit);
+}
+
+CAMLprim value caml_mutex_unlock (value wrapper)
+{
+  CAMLparam1(wrapper);
+  caml_plat_mutex* m = Mutex_val(wrapper);
+  caml_enter_blocking_section();
```
Comment (ctk21) [inline runtime/domain.c]: why do we need the blocking section for unlocking the mutex?
do we need to do the local roots for wrapper?

File: runtime/domain.c [hunk 3]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+  caml_plat_unlock(m);
+  caml_leave_blocking_section();
+  CAMLreturn (Val_unit);
+}
+
+CAMLprim value caml_mutex_try_lock (value wrapper)
+{
+  CAMLparam1(wrapper);
+  caml_plat_mutex* m = Mutex_val(wrapper);
+  caml_enter_blocking_section();
```
Comment (ctk21) [inline runtime/domain.c]: I don't think we need the blocking section or the local roots for try_lock

File: runtime/domain.c [hunk 4]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+  caml_plat_unlock(m);
+  caml_leave_blocking_section();
+  CAMLreturn (Val_unit);
+}
+
+CAMLprim value caml_mutex_try_lock (value wrapper)
```
Comment (ctk21) [inline runtime/domain.c:1554]: This function has a bug. The return value of `caml_plat_try_lock` is not being passed back.

File: runtime/domain.c [hunk 5]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+    caml_alloc_custom(&caml_condition_ops, sizeof(caml_plat_cond*), 0, 1);
+  Condition_val(wrapper) = c;
+  CAMLreturn (wrapper);
+}
+
+CAMLprim value caml_condition_signal (value wrapper)
+{
+  CAMLparam1(wrapper);
+  caml_plat_cond* c = Condition_val(wrapper);
+  caml_enter_blocking_section();
```
Comment (ctk21) [inline runtime/domain.c]: Do we need the blocking section and the local root protection to call the signal?

File: runtime/domain.c [hunk 6]
```diff
@@ -1481,3 +1481,160 @@ CAMLprim value caml_domain_dls_get(value unused)
...
+  caml_plat_signal(c);
+  caml_leave_blocking_section();
+  CAMLreturn (Val_unit);
+}
+
+CAMLprim value caml_condition_broadcast (value wrapper)
+{
+  CAMLparam1(wrapper);
+  caml_plat_cond* c = Condition_val(wrapper);
+  caml_enter_blocking_section();
```
Comment (ctk21) [inline runtime/domain.c]: Do we need the blocking section and the local root protection to call broadcast?

File: stdlib/domain.ml [hunk 7]
```diff
@@ -42,6 +42,22 @@ module Sync = struct
...
+  external unlock : t -> unit = "caml_mutex_unlock"
+  external try_lock : t -> unit = "caml_mutex_try_lock"
+end
+
+module Condition = struct
+  type t
+  external create : Mutex.t -> t = "caml_condition_new"
+  external wait : t -> unit = "caml_condition_wait"
+  external broadcast : t -> unit = "caml_condition_broadcast"
+  external signal : t -> unit = "caml_condition_signal"
```
Comment (ctk21) [inline stdlib/domain.ml]: can `signal` and `broadcast` be `noalloc`?

File: stdlib/domain.ml [hunk 8]
```diff
@@ -42,6 +42,22 @@ module Sync = struct
   external poll : unit -> unit = "%poll"
 end
 
+module Mutex = struct
+  type t
+  external create : unit -> t = "caml_mutex_new"
+  external lock : t -> unit = "caml_mutex_lock"
+  external unlock : t -> unit = "caml_mutex_unlock"
+  external try_lock : t -> unit = "caml_mutex_try_lock"
```
Comment (ctk21) [inline stdlib/domain.ml]: can `try_lock` and `unlock` be `noalloc`

File: stdlib/domain.mli [hunk 9]
```diff
@@ -111,6 +111,22 @@ module Sync : sig
...
 
+module Mutex : sig
+  type t
+  val create : unit -> t
+  val lock : t -> unit
+  val unlock : t -> unit
+  val try_lock : t -> unit
+end
+
+module Condition : sig
```
Comment (ctk21) [inline stdlib/domain.mli:122]: This interface is different for `create` and `wait` vs the systhread `Condition` module. I guess we need to make a call at some point which will be the final one.

**Review by ctk21 — COMMENTED** [review]
I know this PR is a draft, but I will be on vacation for a bit, so figured some feedback now is better than nothing for a while. 

I don't think it matters that this is not unified with the `Mutex` and `Condition` modules as it is in its own namespace. We can experimentally put this in and understand the systhread overhead. 

One thing I'm not sure about is if the symbols (e.g. `caml_mutex_lock`, etc) being the same as those expected by the existing modules (`Mutex` and `Condition`) will cause problems. Particularly for #381 and co-existing with systhreads.

File: runtime/domain.c (hunk 1 above)
Comment (kayceesrk) [reply runtime/domain.c]: done.

File: runtime/domain.c (hunk 2 above)
Comment (kayceesrk) [reply runtime/domain.c]: fixed.

File: runtime/domain.c (hunk 4 above)
Comment (kayceesrk) [reply runtime/domain.c:1554]: fixed.

File: runtime/domain.c (hunk 5 above)
Comment (kayceesrk) [reply runtime/domain.c]: fixed.

File: stdlib/domain.ml (hunk 7 above)
Comment (kayceesrk) [reply stdlib/domain.ml]: yes.

File: runtime/domain.c (hunk 6 above)
Comment (kayceesrk) [reply runtime/domain.c]: fixed.

File: stdlib/domain.ml (hunk 8 above)
Comment (kayceesrk) [reply stdlib/domain.ml]: yes,

**Comment by ctk21** [comment]
Closing this PR as we have implemented and merged it in #548

**Pull request closed without merging**